Basic heap implementation in cpp

dhruvinfo28 opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.
You can learn more about contributing to NeoAlgo here: https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md

Happy Contributing!

-->

### Have you read the [Contributing Guidelines on Pull Requests]

Yes

### Description

closes #2541 . Implemented the basic heap as per the issue.

### Checklist

- [Yes ] I've read the contribution guidelines.
- [ Yes] I've checked the issue list before deciding what to submit.
- [ Yes] I've edited the `README.md` and link to my code.

dhruvinfo28 committed:
Basic heap implementation

HarshCasper reviewed (commented):
Few changes: 
- Add a beginner-friendly description about your Code at the top
- Add description of functions 
- Add complexity at the end of the code
- Use a Code formatter to format the Code

dhruvinfo28 commented:
Sure, I'm on it.

dhruvinfo28 committed:
Requested changes done

Checklist:
Description at the top: yes
description of functions: yes
Complexity added
Also used a code formatter

dhruvinfo28 commented:
Why are checks for other languages failing? 
@HarshCasper @bhav09
